airbyte-workers/src/main/java/io/airbyte/workers/DefaultGetSpecWorker.java

```diff
@@ -41,37 +46,63 @@
   private static final Logger LOGGER = LoggerFactory.getLogger(DefaultGetSpecWorker.class);
 
   private final IntegrationLauncher integrationLauncher;
+  private final AirbyteStreamFactory streamFactory;
 
   private Process process;
 
-  public DefaultGetSpecWorker(final IntegrationLauncher integrationLauncher) {
+  public DefaultGetSpecWorker(final IntegrationLauncher integrationLauncher, AirbyteStreamFactory streamFactory) {
     this.integrationLauncher = integrationLauncher;
+    this.streamFactory = streamFactory;
+  }
+
+  public DefaultGetSpecWorker(final IntegrationLauncher integrationLauncher) {
+    this(integrationLauncher, new DefaultAirbyteStreamFactory());
   }
 
   @Override
   public OutputAndStatus<StandardGetSpecOutput> run(JobGetSpecConfig config, Path jobRoot) {
     try {
       process = integrationLauncher.spec(jobRoot).start();
+
       LineGobbler.gobble(process.getErrorStream(), LOGGER::error);
 
+      Optional<ConnectorSpecification> spec;
       try (InputStream stdout = process.getInputStream()) {
-        // retrieving spec should generally be instantaneous, but since docker images might not be pulled
-        // it could take a while longer depending on internet conditions as well.
-        WorkerUtils.gentleClose(process, 2, TimeUnit.MINUTES);
-
-        if (process.exitValue() == 0) {
-          String specString = new String(stdout.readAllBytes());
-          ConnectorSpecification spec = Jsons.deserialize(specString, ConnectorSpecification.class);
-          return new OutputAndStatus<>(JobStatus.SUCCEEDED, new StandardGetSpecOutput().withSpecification(spec));
-        } else {
+        spec = streamFactory.create(IOs.newBufferedReader(stdout))
+            .filter(message -> message.getType() == Type.SPEC)
+            .map(AirbyteMessage::getSpec)
+            .findFirst();
+      }
+
+      // todo (cgardens) - let's pre-fetch the images outside of the worker so we don't need account for
+      // this.
+      // retrieving spec should generally be instantaneous, but since docker images might not be pulled
+      // it could take a while longer depending on internet conditions as well.
+      WorkerUtils.gentleClose(process, 2, TimeUnit.MINUTES);
+
+      int exitCode = process.exitValue();
+      if (exitCode == 0) {
+        if (spec.isEmpty()) {
+          LOGGER.error("integration failed to output a spec struct.");
           return new OutputAndStatus<>(JobStatus.FAILED);
         }
-      }
 
+        // hack: more json schema X-module issues.
```

Comment:
the ConnectorSpecification is a dependency of two json schema that are in different modules. this means that the pojo generator creates this class twice in different packages. this is pretty horrible, but for we are going to just convert between the two and deal with it.